EoM not reading embeded color profiles.

#### Expected behaviour
Opening an image with a color profile uses the color profile (Adobe RGB, etc.).

#### Actual behaviour

Adobe RBG jpegs look flat, it seems to be reading them as sRGB.

#### Steps to reproduce the behaviour
Open a RAW in Darktable.

Adjust (make sure it has some warm tones in part). Export it once with sRGB, export it once with Adobe RGB.

Open each with EoM, compare to the image you see in Darktable.

#### MATE general version
1.20.0

#### Package version
1.20.0-2ubuntu1

#### Linux Distribution
Ubuntu MATE 18.04

#### Link to downstream report of your Distribution

Tried to build `eom` with LCMS enabled, got the following error:
```
configure: error: Package requirements (gtk+-3.0 >= 3.22.0              glib-2.0 >= 2.50.0              gmodule-2.0              gio-2.0 >= 2.50.0              gio-unix-2.0 >= 2.50.0              gthread-2.0 >= 2.50.0              mate-desktop-2.0 >= 1.17.0              gdk-pixbuf-2.0 >= 2.30.0              gio-2.0 >= 2.50.0              gtk+-unix-print-3.0 >= 3.22.0              shared-mime-info >= 0.20              libpeas-1.0 >= 1.8.0              libpeas-gtk-1.0 >= 1.8.0 lcms2 librsvg-2.0 >= 2.36.2) were not met:

Package 'gtk+-3.0', required by 'virtual:world', not found
Package 'mate-desktop-2.0', required by 'virtual:world', not found
Package 'gtk+-unix-print-3.0', required by 'virtual:world', not found
Package 'libpeas-1.0', required by 'virtual:world', not found
Package 'libpeas-gtk-1.0', required by 'virtual:world', not found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables EOM_CFLAGS
and EOM_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
```
I don't seem to have `mate-desktop` version `2.0` available? :thinking:

--- a/src/eom-image.c
+++ b/src/eom-image.c
@@ -632,10 +632,32 @@ eom_image_apply_display_profile (EomImage *img, cmsHPROFILE screen)
 	if (screen == NULL) return;
 
 	if (priv->profile == NULL) {
-		/* Assume sRGB color space for images without ICC profile */
-		eom_debug_message (DEBUG_LCMS, "Image has no ICC profile. "
-		                   "Assuming sRGB.");
-		priv->profile = cmsCreate_sRGBProfile ();
+		/* Check whether GdkPixbuf was able to extract a profile */
+		const char* data = gdk_pixbuf_get_option (priv->image,
+		                                          "icc-profile");
+
+		if(data) {
+			gsize   profile_size = 0;
+			guchar *profile_data = g_base64_decode(data,
+			                                       &profile_size);
+
+			if (profile_data && profile_size > 0) {
+				eom_debug_message (DEBUG_LCMS,
+				                   "Using ICC profile "
+				                   "extracted by GdkPixbuf");
+				priv->profile =
+					cmsOpenProfileFromMem(profile_data,
+					                      profile_size);
+				g_free(profile_data);
+			}
+		}
+
+		if(priv->profile == NULL) {
+			/* Assume sRGB color space for images without ICC profile */
+			eom_debug_message (DEBUG_LCMS, "Image has no ICC profile. "
+					   "Assuming sRGB.");
+			priv->profile = cmsCreate_sRGBProfile ();
+		}
 	}
 
 	/* TODO: support other colorspaces than RGB */

--- a/src/eom-image.c
+++ b/src/eom-image.c
@@ -629,7 +629,14 @@ eom_image_apply_display_profile (EomImage *img, cmsHPROFILE screen)
 
 	priv = img->priv;
 
-	if (screen == NULL || priv->profile == NULL) return;
+	if (screen == NULL) return;
+
+	if (priv->profile == NULL) {
+		/* Assume sRGB color space for images without ICC profile */
+		eom_debug_message (DEBUG_LCMS, "Image has no ICC profile. "
+		                   "Assuming sRGB.");
+		priv->profile = cmsCreate_sRGBProfile ();
+	}
 
 	/* TODO: support other colorspaces than RGB */
 	if (cmsGetColorSpace (priv->profile) != cmsSigRgbData ||

--- a/src/eom-image.c
+++ b/src/eom-image.c
@@ -667,18 +667,17 @@ eom_image_apply_display_profile (EomImage *img, cmsHPROFILE screen)
 		return;
 	}
 
-	/* TODO: find the right way to colorcorrect RGBA images */
-	if (gdk_pixbuf_get_has_alpha (priv->image)) {
-		eom_debug_message (DEBUG_LCMS, "Colorcorrecting RGBA images is unsupported.");
-		return;
-	}
+	cmsUInt32Number color_type = TYPE_RGB_8;
+
+	if (gdk_pixbuf_get_has_alpha (priv->image))
+		color_type = TYPE_RGBA_8;
 
 	transform = cmsCreateTransform (priv->profile,
-				        TYPE_RGB_8,
-				        screen,
-				        TYPE_RGB_8,
-				        INTENT_PERCEPTUAL,
-				        0);
+	                                color_type,
+	                                screen,
+	                                color_type,
+	                                INTENT_PERCEPTUAL,
+	                                0);
 
 	if (G_LIKELY (transform != NULL)) {
 		rows = gdk_pixbuf_get_height (priv->image);

--- a/src/eom-metadata-reader-png.c
+++ b/src/eom-metadata-reader-png.c
@@ -574,11 +574,11 @@ eom_metadata_reader_png_get_icc_profile (EomMetadataReaderPng *emr)
 		profile = cmsCreate_sRGBProfile ();
 	}
 
-	if (!profile && priv->cHRM_chunk) {
+	if (!profile && priv->cHRM_chunk && priv->gAMA_chunk) {
 		cmsCIExyY whitepoint;
 		cmsCIExyYTRIPLE primaries;
 		cmsToneCurve *gamma[3];
-		double gammaValue = 2.2; // 2.2 should be a sane default gamma
+		double gammaValue;
 
 		/* This uglyness extracts the chromacity and whitepoint values
 		 * from a PNG's cHRM chunk. These can be accurate up to the
@@ -597,14 +597,9 @@ eom_metadata_reader_png_get_icc_profile (EomMetadataReaderPng *emr)
 		primaries.Blue.x = EXTRACT_DOUBLE_UINT_BLOCK_OFFSET (priv->cHRM_chunk, 6, 100000);
 		primaries.Blue.y = EXTRACT_DOUBLE_UINT_BLOCK_OFFSET (priv->cHRM_chunk, 7, 100000);
 
-		primaries.Red.Y = primaries.Green.Y = primaries.Blue.Y = 1.0;
+		whitepoint.Y = primaries.Red.Y = primaries.Green.Y = primaries.Blue.Y = 1.0;
 
-		/* If the gAMA_chunk is present use its value which is saved
-		 * the same way as the whitepoint. Use 2.2 as default value if
-		 * the chunk is not present. */
-		if (priv->gAMA_chunk)
-			gammaValue = (double) 1.0/EXTRACT_DOUBLE_UINT_BLOCK_OFFSET (priv->gAMA_chunk, 0, 100000);
-
+		gammaValue = (double) 1.0/EXTRACT_DOUBLE_UINT_BLOCK_OFFSET (priv->gAMA_chunk, 0, 100000);
 		gamma[0] = gamma[1] = gamma[2] = cmsBuildGamma (NULL, gammaValue);
 
 		profile = cmsCreateRGBProfile (&whitepoint, &primaries, gamma);

--- a/src/eom-metadata-reader-png.c
+++ b/src/eom-metadata-reader-png.c
@@ -25,6 +25,7 @@
 #include <config.h>
 #endif
 
+#include <math.h>
 #include <string.h>
 #include <zlib.h>
 
@@ -489,6 +490,31 @@ eom_metadata_reader_png_get_xmp_data (EomMetadataReaderPng *emr )
  * A maximum output buffer of 5MB should be enough. */
 #define EOM_ICC_INFLATE_BUFFER_LIMIT (1024*1024*5)
 
+/* Apparently an sRGB profile saved in cHRM and gAMA chunks does not compute
+ * a profile that exactly matches the built-in sRGB profile and thus could
+ * cause a slight color deviation. Try catching this case to allow fallback
+ * to the built-in profile instead.
+ */
+static gboolean
+_chrm_matches_srgb(const cmsCIExyY       *whitepoint,
+		   const cmsCIExyYTRIPLE *primaries,
+		         gdouble          gammaValue)
+{
+	/* PNGs gAMA value for 2.2 is only accurate to the 4th decimal point */
+#define DOUBLE_EQUAL_MAX_DIFF 1e-4
+#define DOUBLE_EQUAL(a,b) (fabs (a - b) < DOUBLE_EQUAL_MAX_DIFF)
+
+	return (DOUBLE_EQUAL(gammaValue, 2.2)
+		&& DOUBLE_EQUAL(whitepoint->x, 0.3127)
+		&& DOUBLE_EQUAL(whitepoint->y, 0.329)
+		&& DOUBLE_EQUAL(primaries->Red.x, 0.64)
+		&& DOUBLE_EQUAL(primaries->Red.y, 0.33)
+		&& DOUBLE_EQUAL(primaries->Green.x, 0.3)
+		&& DOUBLE_EQUAL(primaries->Green.y, 0.6)
+		&& DOUBLE_EQUAL(primaries->Blue.x, 0.15)
+		&& DOUBLE_EQUAL(primaries->Blue.y, 0.06));
+}
+
 static gpointer
 eom_metadata_reader_png_get_icc_profile (EomMetadataReaderPng *emr)
 {
@@ -600,11 +626,20 @@ eom_metadata_reader_png_get_icc_profile (EomMetadataReaderPng *emr)
 		whitepoint.Y = primaries.Red.Y = primaries.Green.Y = primaries.Blue.Y = 1.0;
 
 		gammaValue = (double) 1.0/EXTRACT_DOUBLE_UINT_BLOCK_OFFSET (priv->gAMA_chunk, 0, 100000);
-		gamma[0] = gamma[1] = gamma[2] = cmsBuildGamma (NULL, gammaValue);
+		eom_debug_message (DEBUG_LCMS, "Gamma %.5lf", gammaValue);
 
-		profile = cmsCreateRGBProfile (&whitepoint, &primaries, gamma);
-
-		cmsFreeToneCurve(gamma[0]);
+		/* Catch SRGB in cHRM/gAMA chunks and use accurate built-in
+		 * profile instead of computing one that "gets close". */
+		if(_chrm_matches_srgb (&whitepoint, &primaries, gammaValue)) {
+			eom_debug_message (DEBUG_LCMS, "gAMA and cHRM match sRGB");
+			profile = cmsCreate_sRGBProfile ();
+		} else {
+			gamma[0] = gamma[1] = gamma[2] =
+				cmsBuildGamma (NULL, gammaValue);
+			profile = cmsCreateRGBProfile (&whitepoint, &primaries,
+						       gamma);
+			cmsFreeToneCurve(gamma[0]);
+		}
 	}
 
 	return profile;

--- a/src/eom-window.c
+++ b/src/eom-window.c
@@ -388,7 +388,7 @@ eom_window_get_display_profile (GdkScreen *screen)
 	gulong length;
 	guchar *str;
 	int result;
-	cmsHPROFILE *profile;
+	cmsHPROFILE *profile = NULL;
 	char *atom_name;
 
 	dpy = GDK_DISPLAY_XDISPLAY (gdk_screen_get_display (screen));
@@ -440,14 +440,17 @@ eom_window_get_display_profile (GdkScreen *screen)
 
 		if (G_UNLIKELY (profile == NULL)) {
 			eom_debug_message (DEBUG_LCMS,
-					   "Invalid display profile, "
-					   "not correcting");
+					   "Invalid display profile set, "
+					   "not using it");
 		}
 
 		XFree (str);
-	} else {
-		profile = NULL;
-		eom_debug_message (DEBUG_LCMS, "No profile, not correcting");
+	}
+
+	if (profile == NULL) {
+		profile = cmsCreate_sRGBProfile ();
+		eom_debug_message (DEBUG_LCMS,
+				 "No valid display profile set, assuming sRGB");
 	}
 
 	return profile;

